keylime/agentstates.py

```diff
@@ -6,11 +6,14 @@
 
 import threading
 
-from keylime.ima_ast import START_HASH, FF_HASH
+from keylime.ima_ast import get_FF_HASH, get_START_HASH
+from keylime.common.algorithms import Hash
 from keylime.ima_file_signatures import ImaKeyrings
 
 class TPMState():
     """ TPMState models the state of the TPM's PCRs """
+    hash_alg = Hash.SHA256  # TODO: This should be configurable
```

Comment:
I think we need to be able to configure this in AgentAttestStates.add() when calling set_ima_pcrs(). The mess is with overlap of sm3 versus sha256 via the size because at the moment we can only derive usage of the PCR bank/hash via the size of the hash, which is a bit ambiguous. The agent would have to tell use what it is using...

Reply:
>  The agent would have to tell use what it is using...

The problem is that we don't actually know the algorithm before the first attestation round.

What  do you think about just keeping a PCR for every possible hash type that we support and when we update (`update_ima_attestation(..)`) or get PCR values (`get_pcr_state(...)`) you'll have to specify a hash algorithm?

Reply:
> What do you think about just keeping a PCR for every possible hash type that we support and when we update (update_ima_attestation(..)) or get PCR values (get_pcr_state(...)) you'll have to specify a hash algorithm?

I would try to not keep it for ever possible hash. The get_pcr_state() in _process_measurement_list could request PCR 10 for sha1 or sha256 (as examples) by inferring from the size of the pcrval_bytes() which algo is used (let's toss out sm3 support due to ambiguity to infer from size or have the agent say what exactly it uses). But then get_pcr_state() should return any existing value that it has been given (restored from DB) and only ever do a reset of a PCR value if the PCR holds the START_HASH value and adjust it for example from sha1 to sha256. Does this make sense?

Reply:
When we call `get_pcr_state()` in `_process_measurement_list` we actually know the hashing algorithm that was used, so there is no need to infer them. 
We then just need to lazily initialize the `START_FF` hashes because during object creation we don't actually now the hash algorithm.

Reply:
@stefanberger I've now added a implementation for that.